### PR TITLE
Fix serialization of shared pointers

### PR DESCRIPTION
**Pull request by akva2**
commit https://github.com/OPM/opm-common/commit/067d6d44460548e10610b5c753bd2db76bb1887f aimed to
keep the shared nature of shared pointers when deserializing.

the problem is that assigning multiple shared ptrs to the same
raw pointer does not enable any sharing, it only leads to multiple
deletions. fix this by properly holding a shared pointer in the map.

convert the adresses that is serialized to keep track of what pointers were
shared on the sender side to integers (uintptr_t) rather than serializing
them as void pointers. this is clearer and in principle more correct as
it is rather unclear how a serializer should handle a void pointer, and
it only worked by "accident" since std::is_pod_v happens to return true
for a void*.

while at it make sure to clear out pointer maps after pack/unpack operations for
memory savings (does not matter in reality since serializer objects are immediately destroyed,
but still..)

### CODE REVIEW

**Comment by akva2** [comment]
jenkins build this please

File: opm/common/utility/Serializer.hpp [hunk 1]
```diff
@@ -545,53 +549,47 @@ class Serializer {
...
 
-    //! \brief Handler for smart pointers.
+    //! \brief Handler for shared pointers.
     template<class PtrType>
-    void ptr(const PtrType& data)
+    void shared_ptr(const PtrType& data)
     {
         using T1 = typename PtrType::element_type;
-        void* data_ptr = reinterpret_cast<void*>(data.get());
+        uintptr_t data_ptr = reinterpret_cast<uintptr_t>(data.get());
```
Comment (bska) [inline opm/common/utility/Serializer.hpp:557]: Tiny nit: Using `uintptr_t` means you need to `#include <stdint.h>`.  Optionally, you could use `std::uintptr_t` in which case you'd need to `#include <cstdint>`.

File: opm/common/utility/Serializer.hpp [hunk 2]
```diff
@@ -545,53 +549,47 @@ class Serializer {
...
         (*this)(data_ptr);
         if (!data_ptr)
             return;
-
         if (m_op == Operation::PACK || m_op == Operation::PACKSIZE) {
             if (m_ptrmap.count(data_ptr) == 0) {
                 (*this)(*data);
-                m_ptrmap[data_ptr] = nullptr;
+                m_ptrmap[data_ptr] = 0;
```
Comment (bska) [inline opm/common/utility/Serializer.hpp:564]: I think I'd prefer `m_ptrmap[data_ptr].reset()` here, or maybe just `m_ptrmap.insert_or_assign(data_ptr, std::make_shared<void>())`.

**Review by bska — COMMENTED** [review]
Other than some minor nits this looks good to me.

File: opm/common/utility/Serializer.hpp (hunk 1 above)
Comment (akva2) [reply opm/common/utility/Serializer.hpp]: yeah, my bad. used the std:: prefixed one.

File: opm/common/utility/Serializer.hpp (hunk 2 above)
Comment (akva2) [reply opm/common/utility/Serializer.hpp]: i used the first one. would have liked the second one but std::make_shared can't be used for a void pointer

**Review by bska — APPROVED** [review]
Thanks a lot for the updates.  This looks good to me now and I'll merge into master as well as back-port the fix to the 2024.10 release branch.

**Pull request merged**